### PR TITLE
REV-25-fetch agents and pagination and type search

### PR DESCRIPTION
**Pull request by IbrahimAldereni-socuimtech**
### Description
- add endpoint to get and search agent list

- backend repo: https://github.com/abukhadraop-or/reverifi-api/pull/16

- figma: https://www.figma.com/file/22d3bKcFL0ipWIZJOUY3Km/reverifi?node-id=0%3A7349

### CODE REVIEW

File: src/components/agent-list/AgentCardsContainer/agent-cards-container.styles.js [hunk 1]
```diff
@@ -24,3 +34,170 @@ export const CardsContainer = styled.div`
...
+export const HiddenComponent = styled.div`
+  visibility: hidden;
+`;
+
+// export const Container = styled.div`
+//   display: flex;
+//   justify-content: center;
+//   margin: 1rem 0;
+//   width: 100%;
+// `;
```
Comment (AhmadAlShwaiki) [inline src/components/agent-list/AgentCardsContainer/agent-cards-container.styles.js:47]: Please remove any unused code

File: src/components/agent-list/AgentCardsContainer/index.jsx [hunk 2]
```diff
@@ -14,33 +22,74 @@ import {
...
+
+  /**
+   * handle search function
+   *
+   * @param {Object} target form target object
+   */
+  const handleSearch = ({ target }) => {
+    const { type, location, name } = target;
+    navigate(
+      `?type=${type.value}&location=${location.value}&name=${name.value}`
```
Comment (AhmadAlShwaiki) [inline src/components/agent-list/AgentCardsContainer/index.jsx:67]: Please stick to object destructing, create a function that does the above logic to generate a string from three values as well.

File: src/hooks/use-on-screen.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ * @return {Boolean}  is component in view
+ */
+function useOnScreen(ref) {
+  const [isIntersecting, setIntersecting] = useState(false);
+
+  const observer = new IntersectionObserver(([entry]) =>
+    setIntersecting(entry.isIntersecting)
+  );
+
+  useEffect(() => {
```
Comment (AhmadAlShwaiki) [inline src/hooks/use-on-screen.js:16]: Please use useEffectOnce

File: src/hooks/use-on-screen.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import { useEffect, useState } from 'react';
+
+/**
+ * hook that check if a component is on view
+ *
+ * @param {Ref} ref component reference
+ * @return {Boolean}  is component in view
+ */
+function useOnScreen(ref) {
+  const [isIntersecting, setIntersecting] = useState(false);
```
Comment (AhmadAlShwaiki) [inline src/hooks/use-on-screen.js:10]: Please stick to the best practices for function naming; This is pretty essential 
```suggestion
  const [isIntersecting, setIsintersecting] = useState(false);
```

File: src/services/user.js [hunk 5]
```diff
@@ -16,6 +16,27 @@ export const updateUserRoles = async (rolesIds) => {
...
+ * @param {number} [page=0]
+ *
+ * @return {Object[]} Array of agents.
+ */
+export const getUsersByType = async (type, location, name, page = 0) => {
+  const userType = type.charAt(0).toUpperCase() + type.slice(1);
+  const {
+    data: { data },
+  } = await http.get(
+    `${apiUrl}/users/roles/${userType}?page=${page}&location=${location}&name=${name}`
```
Comment (AhmadAlShwaiki) [inline src/services/user.js:34]: Same as above for generating url strings

File: src/services/user.js [hunk 6]
```diff
@@ -16,6 +16,27 @@ export const updateUserRoles = async (rolesIds) => {
...
+ *
+ * @param {String} type agent type
+ * @param {String} location agent location
+ * @param {String} name agent name
+ * @param {number} [page=0]
+ *
+ * @return {Object[]} Array of agents.
+ */
+export const getUsersByType = async (type, location, name, page = 0) => {
+  const userType = type.charAt(0).toUpperCase() + type.slice(1);
```
Comment (AhmadAlShwaiki) [inline src/services/user.js:30]: Please create a util function for this one.

File: src/components/agent-list/AgentCardsContainer/agent-cards-container.styles.js [hunk 7]
```diff
@@ -24,3 +34,170 @@ export const CardsContainer = styled.div`
...
+  border-left: 0.0625rem solid ${colors.midGray};
+  cursor: text;
+  font-size: 0.75rem;
+  height: 100%;
+  margin: 0 0.3125rem;
+  padding: 0 0 0 1.5625rem;
+  width: 100%;
+
+  ${mq.mobileWide`
+    font-size: 0.9375rem;
```
Comment (AhmadAlShwaiki) [inline src/components/agent-list/AgentCardsContainer/agent-cards-container.styles.js:168]: Please adhere to the proper css styling. 
```suggestion
    font-size:1rem;
```

File: src/components/agent-list/AgentCardsContainer/index.jsx (hunk 2 above)
Comment (IbrahimAldereni-socuimtech) [reply src/components/agent-list/AgentCardsContainer/index.jsx:67]: I don't understand what you mean

File: src/components/agent-list/AgentCard/index.jsx [hunk 8]
```diff
@@ -101,7 +101,9 @@ function AgentCard({
               </PropertyIconContainer>
               <PropertyText>Address</PropertyText>
             </ContactProperty>
-            <ContactText>{address}</ContactText>
+            <ContactText>
+              {address?.zipCode} {address?.city}, {address?.country}
```
Comment (AhmadAlShwaiki) [inline src/components/agent-list/AgentCard/index.jsx:105]:
```suggestion
              {getContactList(address)}
```

File: src/components/agent-list/AgentCardsContainer/agent-cards-container.styles.js [hunk 9]
```diff
@@ -24,3 +34,163 @@ export const CardsContainer = styled.div`
...
+  border-radius: 1.75rem;
+  box-shadow: 0 0.0625rem 0.3rem ${colors.midGray};
+  color: ${colors.gray};
+  display: flex;
+  justify-content: space-between;
+  margin: 0.125rem 0.3rem;
+  padding: 0.3rem;
+`;
+
+export const Heading = styled.p`
```
Comment (AhmadAlShwaiki) [inline src/components/agent-list/AgentCardsContainer/agent-cards-container.styles.js:65]: This component is seen elsewhere, can we reuse it?

File: src/components/agent-list/AgentCardsContainer/agent-cards-container.styles.js [hunk 10]
```diff
@@ -24,3 +34,163 @@ export const CardsContainer = styled.div`
...
+  justify-content: space-between;
+  width: 100%;
+`;
+
+export const FilterProperty = styled.div`
+  align-items: center;
+  display: flex;
+
+  ${mq.tablet`
+    display: none;
```
Comment (AhmadAlShwaiki) [inline src/components/agent-list/AgentCardsContainer/agent-cards-container.styles.js:83]: 👍

File: src/components/agent-list/AgentCardsContainer/agent-cards-container.styles.js [hunk 11]
```diff
@@ -24,3 +34,163 @@ export const CardsContainer = styled.div`
...
+`;
+
+export const AgentSearch = styled.input`
+  border: none;
+  border-left: 0.0625rem solid ${colors.midGray};
+  cursor: text;
+  font-size: 0.75rem;
+  height: 100%;
+  margin: 0 0.3rem;
+  padding: 0 0 0 1.6rem;
```
Comment (AhmadAlShwaiki) [inline src/components/agent-list/AgentCardsContainer/agent-cards-container.styles.js:157]:
```suggestion
  padding-left: 1.6rem;
```

File: src/components/agent-list/AgentCardsContainer/agent-cards-container.styles.js [hunk 12]
```diff
@@ -24,3 +34,163 @@ export const CardsContainer = styled.div`
...
+  width: 100%;
+
+  ${mq.mobileWide`
+    font-size: 1rem;
+  `};
+
+  ${mq.tablet`
+    display: flex;
+    padding: 0.625rem;
```
Comment (AhmadAlShwaiki) [inline src/components/agent-list/AgentCardsContainer/agent-cards-container.styles.js:166]: Please cleanup magical css values by rounding them.

File: src/hooks/use-on-screen.js [hunk 13]
```diff
@@ -0,0 +1,28 @@
+import { useState } from 'react';
+import useEffectOnce from './use-effect-once';
+
+/**
+ * hook that check if a component is on view
+ *
+ * @param {Ref} ref component reference
+ * @return {Boolean}  is component in view
+ */
+function useOnScreen(ref) {
```
Comment (AhmadAlShwaiki) [inline src/hooks/use-on-screen.js:10]: Cant we just replace this hook with a pagination component that takes props from it's parent and renders them conditionally?
```
export default function Pagination({ nextPage, onChange, currentPage, prevPage, updatePage }) {
  return (
    <>
      {prevLink && (
        <Button
          onClick={() => {
            if (page > 1) onPageChange(page - 1, updatePage);
          }}
        >
          <IconLeft />
        </Button>
      )}
      <PageCount>{`${t('page')} ${page}`}</PageCount>
      {nextLink && (
        <Button
          onClick={() => {
            onPageChange(page + 1, updatePage);
          }}
        >
          <IconRight />
        </Button>
      )}
</>
  );
}
```

File: src/utils/helpers.js [hunk 14]
```diff
@@ -14,4 +15,12 @@ const getDatesDifference = (date, period) => {
   return diffInDays < period;
 };
 
-export default getDatesDifference;
+/**
```
Comment (AhmadAlShwaiki) [inline src/utils/helpers.js:18]:
```suggestion
toUpperCaseFirstLetter
/**
```

File: src/utils/helpers.js [hunk 15]
```diff
@@ -14,4 +15,12 @@ const getDatesDifference = (date, period) => {
...
+ * capitalize First Letter function
+ *
+ * @param {String} text text to be capitalize first letter
+ *
+ * @return {String} capitalized string
+ */
+export const capitalizeFirstLetter = (text) =>
+  text.charAt(0).toUpperCase() + text.slice(1);
```
Comment (AhmadAlShwaiki) [inline src/utils/helpers.js:26]:
```suggestion
// add string check for the param type.
  text.charAt(0).toUpperCase() + text.slice(1);
```

File: src/components/agent-list/AgentCardsContainer/agent-cards-container.styles.js (hunk 9 above)
Comment (IbrahimAldereni-socuimtech) [reply src/components/agent-list/AgentCardsContainer/agent-cards-container.styles.js:65]: I can't see where is it used??

**Comment by AhmadAlShwaiki** [comment]
visiting `http://reverifi.docker/agent-list` would query for page 1 and 2, this needs to be fixed.

**Review by AhmadAlShwaiki — APPROVED** [review]

**Pull request merged**